docs/RFCS/20181227_follower_reads_implementation.md

```diff
@@ -0,0 +1,298 @@
+the binPacking policy for all others. This `FollowerReadsAwarePolicy` will check
+to see if the OrigTimestamp of a Txn is before now less `recentDuration` plus a
+clock uncertainty duration. The `followerreadsccl` code will then register this
+new policy and set it to the global var `distsqlplan.ReplicaOraclePolicy`.
+
+### Expose DistSender For Determining Follower Read Safety
+
+The last hurdle to exposing follower reads is that the `kv.DistSender` attempts
+to send batch requests to current lease holders which may prevent reads from
+going to nearby follower replicas. In order to inform the DistSender that it can
+send a batch request to a follower we add a new global var in the kv package
+
+```go
+var CanSendToFollower = func(
+   clusterID uuid.UUID, _ *cluster.Settings, _ *roachpb.BatchRequest,
+) bool {
+   return false
+}
+```
+
+Which is adopted by the DistSender when it decides whether to look up a cached
+lease holder in `DistSender.sendSingleRange`. The followerreadsccl package can
+then inject a new implementation of this function which ensures that the batch
+request is a read only transaction and then verifies that it meets the criteria
+for a follower read.
+
+## Drawbacks
+
+There are very few drawbacks to implementing the high level idea.  Most of the
+groundwork has already been laid. Any obvious downsides come from the impurity
+of the injection required to realize the functionality as an enterprise feature.
+
+## Rationale and Alternatives
+
+### Stateful closed timestamp tracking
+
+One potential downside of this approach is that in an edge case it may have the
+potential to have a detrimentally effect cluster performance in the case of
+bursty traffic and a large volume of follower reads. Imagine a situation where a
+large percentage of client traffic is due to follower reads and the cluster is
+heavily loaded such that all transactions are performing acceptably but if the
+workload were to be shifted entirely such that all requests were forced to go to
+leaseholders it would not be capable of acceptably serving the traffic. If then,
+a burst of load or some other cluster event were to lead one or more replicas to
+fall behind in its ability to publish closed timestamps, all traffic which was
+spread over all of the replicas would begin recieving all of the load that had
+been going to followers. It is possible that this concern is not realistic in
+most common cases.  Furthermore it seems straightforward to mitigate by
+increasing the target multiple. The problem seems worse as the replication
+factor increases beyond 3 to numbers like 7 or 9. Furthermore even if the
+increased load during this bursty period does not meaningfully affect OLTP
+traffic, it may lead to potentially massively increased latency for queries
+which in the previous regime had been fast.
+
+A more sophisticated mechanism which statefully tracks a closed timestamps on a
```

Comment:
This is roughly where I was going above, except that I don't think you'd need to track individual replicas (which I agree is not a good thing to buy into). Looking at the node-wide closed timestamp alone should be sufficient, at least for now.